### PR TITLE
New libraries

### PR DESCRIPTION
**Pull request by PhiLhoSoft**
Added cssobj and j2c.
I will stop adding libraries here, I think we covered most libraries of this kind (view-agnostic, server & client, etc.).
I will concentrate now on some other points: mixins, auto-prefixing, etc.

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-    "presets": ["node6", "react"]
+    "presets": ["node6"]
```
Comment (PepijnSenders) [inline .babelrc]: 👍

**Comment by PepijnSenders** [comment]
Nice libraries @PhiLhoSoft, LGTM!

**Comment by PhiLhoSoft** [comment]
Of course, you will need to update the readme with the new stats, when you have some time. 😄

**Comment by pygy** [comment]
Hello folks, `j2c` author here. 

Thanks for adding it and glad to see it performs well among the libs that preserve the JS source => CSS order.

The v1 branch (currently in pre-release at `1.0.0-3`) is even faster (consistent third place), and adds per-instance mangled suffixes rather than per `sheet()` call. It is release-ready but for documentation.

Side note: @PhiLhoSoft would you perchance be a Lua-L regular?

File: src/classes-overload-test/cases/j2c.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import j2c from 'j2c';
+
+import { stylesheet, buttonClassNames } from '../styles';
+import { renderHtml, renderBody } from '../render';
+import { toClasses } from '../../utilities';
+
+export const j2cCase = (caseName) => {
+    const css = j2c.sheet(toClasses(stylesheet));
```
Comment (pygy) [inline src/classes-overload-test/cases/j2c.js]: Would you mind hoisting the `toClasses()` call out of the benchmarked loop?

File: src/nested-test/cases/j2c.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import j2c from 'j2c';
+
+import { createStylesheet } from '../styles';
+import { renderHtml, renderBody } from '../render';
+import { toClasses } from '../../utilities';
+
+export const j2cCase = (caseName) => {
+    const options = { prefixPseudo: true };
```
Comment (pygy) [inline src/nested-test/cases/j2c.js]: Likewise, here, `&` is optional in `j2c` and is a bit slower to process.

`{'.foo': {"[bar='Tom & Jerry']":{baz: 5}}}` should not put `.foo` between Tom and Jerry, so `j2c` lexes the sub-selector when it finds `&`.

I'll send PRs if you don' t mind...

File: src/classes-overload-test/cases/j2c.js (hunk 2 above)
Comment (pygy) [inline src/classes-overload-test/cases/j2c.js]: `j2c` has a different syntax. It supports arbitrary selectors, so dots are mandatory to denote classes. In real code, you don't use a helper function to add the dots at run time, you put them in the source...

Comment (PhiLhoSoft) [inline src/classes-overload-test/cases/j2c.js]: No, it looks like a good suggestion, this call is made only to adapt test data to the library, it is not done in a real project, of course. I should even externalize the calls to createStyleSheet, it is a constant of each test case.

Comment (PepijnSenders) [inline src/classes-overload-test/cases/j2c.js]: No no, all these tests are about generating classNames with their styles, if you will remove the sheet, there is nothing to test anymore. The test is about a flow of operations you would do on each for example page load. So it has to go from start to end.

File: src/nested-test/cases/j2c.js (hunk 3 above)
Comment (PhiLhoSoft) [inline src/nested-test/cases/j2c.js]: prefixPseudo is, as the name implies, only for pseudo-classes: some libraries, including your, I think, need to write `'&:hover'` to get the pseudo-class to stick to the class name. Others handle this automatically. But as said above, this must be done outside the test loop. Will do. (Not sure if the impact is significant, but worth a try, for fairness.)

**Comment by PhiLhoSoft** [comment]
> Side note: @PhiLhoSoft would you perchance be a Lua-L regular?

Yes 😄 That's how I noticed your library, initially... Although I am a bit away from Lua these days.

File: src/classes-overload-test/cases/j2c.js (hunk 2 above)
Comment (PepijnSenders) [inline src/classes-overload-test/cases/j2c.js]: Yeah I know :), but I think even hoisting toClasses out, will give this library an unfair benefit. Or maybe it is not even suited to be benchmarked against the other libraries.

File: src/nested-test/cases/j2c.js (hunk 3 above)
Comment (pygy) [inline src/nested-test/cases/j2c.js]: The `&` is optional for mere concatenation. You'll need for more complex cases like 
`{'.foo': {'& > &': ...}}` or `{'.foo': {'body:global(.firefox) &': ...}}`

**Comment by pygy** [comment]
v1 allows to create distinct instances of `j2c` and each instance gets a random but stable suffix. 

```JS
css = j2c.sheet({'.foo': {bar: 5})
css2 = j2c.sheet({'.foo': {bar: 5})
css === css2 // true
styler.names.foo === '.foo' + styler.suffix // true

styler = j2c() // create a new instance
css = styler.sheet({'.foo': {bar: 5})
styler.names.foo === '.foo' + styler.suffix // true
```

They are long by design (I want to avoid collisions, by default, and since they are identical, they gzip well), but they can be truncated (`j2c.suffix = j2c.suffix.slice(0,8)`) or even replaced manually if you want the suffixes to be shared on both ends of the network. Inserting styles on both ends would require framework-dependent plugins that don't exist yet.

IIUC, using a hash of declarations to create the class name means that you can't take advantage of client information to generate custom styles (in an isomorphic scenario). Or maybe they have a workaround? No idea, I'm not familiar enough with the libs that take that approach. At the very least, the styles have to be updated at hydration time. I see also `Fela` and `StyleTron` who take an interesting approach regarding deduplication, but I wonder how you can debug the result.

I could add a post-processing filter to do hashing and deduplication, but it would limit the kind of styles you can express. Currently `.foo > input[type=button]` is localized to `.foo_j2c-xxxx input[type=button]`, you can't do that with most CSS in JS tools.

**Comment by PhiLhoSoft** [comment]
FYI, I "Moved CSS generation / alteration out of the testing loop"
https://github.com/PhiLhoSoft/css-in-js-perf-tests/commit/736b42bef3417b7f81471f0f21d3ddbe64653476
In a branch where I work on the complex test case.
Needs some work before I make a pull request.
I see no real difference in perf, because of variations between runs...
But that's cleaner anyway.

**Pull request merged**